spec-0007/transition_to_rng.py

```diff
@@ -0,0 +1,144 @@
+    `old_name='random_state'` either by keyword or as a positional argument at
+    `position_name=1`. At the time of application, the name of the argument in the
+    function signature is manually changed to the new name, `rng`. If positional
+    use was allowed before, this is not changed.*
+
+    - If the function is called with both `random_state` and `rng`, the decorator
+      raises an error.
+    - If `random_state` is provided as a keyword argument, the decorator passes
+      `random_state` to the function's `rng` argument as a keyword, but the decorator
+      will emit a `DeprecationWarning` about the deprecation of keyword `random_state`.
+    - If `random_state` is provided as a positional argument, the decorator passes
+      `random_state` to the function's `rng` argument by position, but the decorator
+      will emit a `FutureWarning` about the changing behavior of the argument.
+    - If `rng` is provided as a keyword argument, the decorator validates `rng` using
+      `numpy.random.default_rng` before passing it to the function.
+    - If neither `random_state` nor `rng` is provided, the decorator checks whether
+      `np.random.seed` has been used to set the global seed. If so, it emits a
+      `FutureWarning` against use of `numpy.random.seed`. Either way, the decorator
+      calls the function without explicitly passing the `rng` argument.
+
+    To avoid warnings, a user must pass `rng` as a keyword or pass a `Generator`
+    object or `None` by position.
+
+    After the deprecation period, the decorator can be removed, and the function
+    can begin to validate the `rng` argument by calling `np.random.default_rng(rng)`
+    internally.
+
+    * Even though a `FutureWarning` is emitted when the PRNG argument is used
+      by position, the Hinsen principle is violated unless positional use is
+      deprecated.
+
+      - If `None` is passed by position, the function will change from being
+        seeded to being unseeded
+      - If an integer is passed by position, the random stream will change
+
+      This SPEC does not specify whether projects should deprecate positional
+      use or not; we just note that this decorator does *not* deprecate
+      positional use and what the implications are.
+
+    Parameters
+    ----------
+    old_name : str
+        The old name of the PRNG argument (e.g. `seed` or `random_state`).
+    position_num : int, optional
+        The (0-indexed) position of the old PRNG argument (if accepted by position).
+        Maintainers are welcome to eliminate this argument and use, for example,
+        `inspect`, if preferred.
+    dep_version : str, optional
+        The full version number of the library in which warnings about use of the
+        old PRNG argument begin to be emitted. The version in which the behavior
+        will change is assumed to be two versions later.
+
+    """
+    new_name = "rng"
```

Comment:
If this must be a variable, I think it should be all caps to indicate that it is a constant.
Note that it is a pain to use this in the warning strings. Where we would just write:
```python3
`rng`
```
in the string, we need to make it an f-string and write:
```python3
`{new_name}`
```
which takes up a lot of the remaining line width.

I forgot the backticks in several places; please add those before merging.

Reply:
Totally fine to just use this as a literal, I think.